lib/convex/context.ex

```diff
@@ -0,0 +1,1942 @@
+  be returned as a result.
+  """
+  @callback pipeline_performed(options :: Keyword.t, context :: This.t,
+                               state :: any)
+    :: any
+
+  @doc """
+  Same as callback `pipeline_performed/2` but errors will be raised
+  as exceptions and result will be returned directly.
+  """
+  @callback pipeline_performed!(options :: Keyword.t, context :: This.t,
+                                state :: any)
+    :: any
+
+
+  #===========================================================================
+  # API Functions
+  #===========================================================================
+
+  @spec new(callback_module :: module) :: context :: This.t
+  @spec new(callback_module :: module, options :: Keyword.t) :: context :: This.t
+  @doc """
+  Creates a new context using the given callback module.
+
+  Callback module could be one of `Convex.Context.Async`, `Convex.Context.Sync`,
+  `Convex.Context.Process` or any custom module implementing `Convex.Context`
+  behaviour.
+
+  Options `director` and `tags` will be handled by the context, any other
+  options will be passed to the callback module `init/1` function.
+
+    - `director`:
+
+      The director module to use to route operations (See `Convex.Director`).
+      If not specified the director will be retrieved from conviguration
+      (See `Convex.Config`).
+
+    - `tags`: `MapSet` or `list` of tags associated with the created context.
+
+  """
+
+  def new(mod, opts \\ []) when is_atom(mod) do
+    {dir, opts} = extract_opt(opts, :director, &Config.director/0)
+    {tags, opts} = extract_opt(opts, :tags, [])
+    {sub, assigns} = _mod_init(mod, %{}, opts)
+    %Ctx{state: :standby, dir: dir, mod: mod, sub: sub,
+         assigns: assigns, tags: MapSet.new(tags)}
+      |> _update_ident()
+  end
+
+
+  @spec authenticated?(context :: This.t) :: boolean
+  @doc """
+  Tells if the contrext is authenticated.
+  """
+
+  def authenticated?(%Ctx{} = ctx), do: ctx.auth != nil
+
+
+  @spec attached?(context :: This.t) :: boolean
+  @doc """
+  Tells if the contrext is attached to a session.
+  """
+
+  def attached?(%Ctx{} = ctx), do: ctx.sess != nil
+
+
+  @spec operation(context :: This.t) :: operation :: This.op
+  @doc """
+  Returns the context current operation name.
+  """
+
+  def operation(ctx), do: _curr_op_name(ctx)
+
+
+  @spec arguments(context :: This.t) :: args :: map
+  @doc """
+  Returns the context current operation arguments.
+  """
+
+  def arguments(ctx), do: _curr_op_args(ctx)
+
+
+  @spec value(context :: This.t) :: any
+  @doc """
+  Returns the context previous operation result.
+  """
+
+  def value(ctx), do: ctx.result
+
+
+  @spec recast(context :: This.t, callback_module :: module) :: context :: This.t
+  @spec recast(context :: This.t, callback_module :: module, Keyword.t)
+    :: context :: This.t
+  @doc """
+  Create a new context with the specified callback module but keeping most
+  of the current meta-data.
```

Comment:
"most of the current meta-data" is a little vague - as a user I'd like to know what is kept, what is discarded and why.

Reply:
I though I explained what meta-data was extensively in the module documentation an in the next lines I explain how some of the metadata may not be brought over. Because the behaviour depends on which callback module is involved I am not sure how I could be more specific here...